### DISCUSSION TITLE
system identification using sparse dataset

### DISCUSSION THREAD

**Discussion by amirali78frz** [casadi/casadi#3313] [Q&A]
Hi,
I am new to Casadi and I want to know if I can solve my problem using it or not. 
The problem is: 
$\min (\sum_{i=1,\dots,T_i} ||\bar{y_1}(t_{1,i})-y_1(t_{1,i})|| + \sum_{j=1,\dots,T_j} ||\bar{y_2}(t_{2,j})-y_2(t_{2,j})||)$
s.t.   $\dot{x}(t) = f(x(t),u(t),p)$
$\quad \quad       y(t) = h(x(t),u(t))$

which x(t) is my states (5 $\times$ 1). u(t) is the controller input (2 $\times$ 1). y(t) is the output (2 $\times$ 1). p is the unknown set of system parameters (10 $\times$ 1). The goal here is system identification and obtaining the parameters to have a model of the system. I have a dataset which contains samples for $\bar{y_1}$ and $\bar{y_2}$ which are in different time instances. So, I had to define my objective function this way. A more specific definition of the system is written below.( I can obtain yhu, Gh, and Gp for all time instances. $c_{ij}$, $cu$, and $cw$ are the parameters. u and w are the controller inputs.)

    u = (yhu - Gh)^(3/2);
    w = (x1 - Gp)^(3/2);

    xdot(1) = cw*w + c12 * (x2 - x1);
    xdot(2) = c21 * (x1 - x2) + c23 * (x3 - x2);
    xdot(3) = c32 * (x2 - x3) + c34 * (x4 - x3);
    xdot(4) = c43 * (x3 - x4) + c45 * (x5 - x4);
    xdot(5) = c54 * (x4 - x5) - cu*u;

    y = [x1; x5];

Also, I can't use interpolation on the dataset and the outputs should be drawn from the above dynamic. Is it possible to solve this using Casadi? Also, if you know any other way to solve this problem, please tell me.

**Comment by jgillis** [comment]
The answer is yes. For example, do single shooting on a nonregular grid. First step is to construct a CasADI Function that performs integration on an arbitrary sized interval.

**Reply by amirali78frz** [reply]
Thanks for your answer. Do you have or know any document or example I read to see how can I do what you said ?!

**Reply by brunoml5** [reply]
Hi! 

If you are very new to optimization for nonlinear models, maybe use some toolbox based on Casadi which implements single-shooting, collocation, etc. Try these ones:

- [MPCTools](https://bitbucket.org/rawlings-group/octave-mpctools/src/master/): For matlab/octave and python. Use the moving horizon functionality for your problem. From Rawlings' group.
- [do-mpc](https://www.do-mpc.com/en/latest/): For python, mature and full of features. From Sergio Lucia group.
- [rockit](https://gitlab.kuleuven.be/meco-software/rockit): From Joris Gillis. For Python. Built on top of the [opti](https://web.casadi.org/docs/#document-opti) interface from casadi. Easy to use with simple syntax.

If you need the theory, I like this [course](https://www.syscop.de/teaching/ss2023/numerical-optimal-control) from Moritz Diehl and his [book draft](https://www.syscop.de/files/2020ss/NOC/book-NOCSE.pdf) on Numerical Optimal Control.

I hope it helps!

**Reply by amirali78frz** [reply]
Thanks. I think these will help me a lot.
